src/object.c

```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    quicklistSetOptions(lobj->ptr, server.list_max_ziplist_size, server.list_compress_depth);
+    llen = listTypeLength(o);
+    listTypeIterator *iter = listTypeInitIterator(o, start, LIST_TAIL);
+    while (llen--) {
+        listTypeEntry entry;
+        listTypeNext(iter, &entry);
+        quicklistEntry *qe = &entry.entry;
+        if (qe->value) {
+            robj *obj = createObject(OBJ_STRING, sdsnewlen((const char *)qe->value, qe->sz));
+            listTypePush(lobj, obj, LIST_TAIL);
+        } else {
+            ll2string(buf, 64, qe->longval);
+            robj *obj = createObject(OBJ_STRING, sdsnewlen((const char *)buf, strlen(buf)));
+            listTypePush(lobj, obj, LIST_TAIL);
+        }
+    }
+    listTypeReleaseIterator(iter);
+    return lobj;
+}
+/* This is a helper function for the COPY command.
+ * Duplicate a set object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupSetObject(robj *o) {
+    robj *set;
+    setTypeIterator *si;
+    sds elesds;
+    int64_t intobj;
+    int encoding;
+
+    serverAssert(o->type == OBJ_SET);
+
+    /* Create a new set object that have the same encoding as the original object's encoding */
+    switch (o->encoding) {
+        case OBJ_ENCODING_INTSET:
+            set = createIntsetObject();
+            break;
+        case OBJ_ENCODING_HT:
+            set = createSetObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    si = setTypeInitIterator(o);
+    while ((encoding = setTypeNext(si, &elesds, &intobj)) != -1) {
+        if (encoding == OBJ_ENCODING_INTSET) {
+            elesds = sdsfromlonglong(intobj);
+            setTypeAdd(set, elesds);
+            sdsfree(elesds);
+        } else {
+            setTypeAdd(set, elesds);
+        }
+    }
+    setTypeReleaseIterator(si);
+    return set;
+}
+
+/* This is a helper function for the COPY command.
+ * Duplicate a zset object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupZsetObject(robj *o) {
+    robj *zobj;
+    char buf[64];
+    long llen, start = 0;
+    int retflags = ZADD_NONE;
+
+    serverAssert(o->type == OBJ_ZSET);
+    
+    /* Create a new zset object that have the same encoding as the original object's encoding */
+    switch (o->encoding) {
+        case OBJ_ENCODING_ZIPLIST:
+            zobj = createZsetZiplistObject();
+            break;
+        case OBJ_ENCODING_SKIPLIST:
+            zobj = createZsetObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    llen = zsetLength(o);
+    if (zobj->encoding == OBJ_ENCODING_ZIPLIST) {
```

Comment:
Same comment as before, ziplists are binary blocks in an allocation, you can just copy the whole thing.